### PR TITLE
Modify rule S2083(JS): Update text to Education Framework structure (APPSEC-201)

### PR DESCRIPTION
**Pull request by loris-s-sonarsource**

### CODE REVIEW

File: rules/S2083/common/fix/function-based-validation.adoc [hunk 1]
```diff
@@ -0,0 +1,14 @@
+==== Use functions that combine feature and validation
+
+The cleanest way to fix this vulnerability is to use a function that combines:
+
+* the retrieval of a file
+* restrict the retrieval of a file to a specific folder
+
```
Comment (loris-s-sonarsource) [inline rules/S2083/common/fix/function-based-validation.adoc:7]:
```suggestion
```

File: rules/S2083/javascript/how-to-fix-it/node.adoc [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+----
+
+==== Compliant solution
+
+[source,javascript,diff-id=1,diff-type=compliant]
+----
+const fs = require('fs');
+
+function (req, res) {
+  const userFilename = fs.realPath(__dirname + req.query.filename);
```
Comment (gaetan-ferry-sonarsource) [inline rules/S2083/javascript/how-to-fix-it/node.adoc:26]: I would also use path.join and avoid the `__dirname` here.

File: rules/S2083/javascript/how-to-fix-it/node.adoc [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+  const targetDirectory = __dirname + '/notsecret/';
```
Comment (gaetan-ferry-sonarsource) [inline rules/S2083/javascript/how-to-fix-it/node.adoc:28]: And here

File: rules/S2083/javascript/how-to-fix-it/node.adoc [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+  if (!realPath.startsWith(targetDirectory)) {
```
Comment (gaetan-ferry-sonarsource) [inline rules/S2083/javascript/how-to-fix-it/node.adoc:29]: I don't think `realPath` is defined here.

File: rules/S2083/javascript/how-to-fix-it/node.adoc [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+
+[source, javascript]
+----
+const fs = require('fs');
+
+function (req, res) {
+  const userFilename = fs.realPath(__dirname + req.query.filename);
+
+  const targetDirectory = __dirname + '/User/John';
```
Comment (gaetan-ferry-sonarsource) [inline rules/S2083/javascript/how-to-fix-it/node.adoc:56]: Use `path.join` here. Also I would avoid `__dirname`

**Review by gaetan-ferry-sonarsource — CHANGES_REQUESTED** [review]
Left a few command. I think the code samples can be improved a little. Also one or two suggestions regarding pitfalls.

File: rules/S2083/javascript/how-to-fix-it/express.adoc [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+using untrusted data. +
+This path is then used without validation to return a file.
+
+Therefore, it can be leveraged by an attacker to delete arbitrary files.
+
+==== Non-compliant code example
+
+[source,javascript,diff-id=1,diff-type=noncompliant]
+----
+const fs = require('fs');
```
Comment (loris-s-sonarsource) [inline rules/S2083/javascript/how-to-fix-it/express.adoc:13]:
```suggestion
const fs = require('fs');
const path = require('path');
```

File: rules/S2083/javascript/how-to-fix-it/express.adoc [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+  res.sendFile(userFilename); // Noncompliant
+}
+----
+
+==== Compliant solution
+
+[source,javascript,diff-id=1,diff-type=compliant]
+----
+const fs = require('fs');
```
Comment (loris-s-sonarsource) [inline rules/S2083/javascript/how-to-fix-it/express.adoc:26]:
```suggestion
const fs = require('fs');
const path = require('path');
```

File: rules/S2083/javascript/how-to-fix-it/node.adoc [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+
+  let data = fs.readFileSync(userFilename, { encoding: 'utf8', flag: 'r' }); // Noncompliant
+}
+----
+
+==== Compliant solution
+
+[source,javascript,diff-id=1,diff-type=compliant]
+----
+const fs = require('fs');
```
Comment (loris-s-sonarsource) [inline rules/S2083/javascript/how-to-fix-it/node.adoc:23]:
```suggestion
const fs = require('fs');
const path = require('path');
```

File: rules/S2083/javascript/how-to-fix-it/node.adoc (hunk 4 above)
Comment (loris-s-sonarsource) [reply rules/S2083/javascript/how-to-fix-it/node.adoc:29]:
```suggestion
  if (!userFilename.startsWith(resourcesDir)) {
```

File: rules/S2083/javascript/how-to-fix-it/node.adoc (hunk 2 above)
Comment (loris-s-sonarsource) [reply rules/S2083/javascript/how-to-fix-it/node.adoc:26]:
```suggestion
  const resourcesDir = "/data/app/resources/"
  const userFilename = path.join(resourcesDir, req.query.filename));
  const userFilename = fs.realPath(userFilename);
```

File: rules/S2083/javascript/how-to-fix-it/node.adoc [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+
+include::../../common/pitfalls/partial-path-traversal.adoc[]
+
+For example, the following code is vulnerable to partial path injection. Note
+that the string `targetDirectory` does not end with a path separator:
+
+
+[source, javascript]
+----
+const fs = require('fs');
```
Comment (loris-s-sonarsource) [inline rules/S2083/javascript/how-to-fix-it/node.adoc:51]:
```suggestion
const fs = require('fs');
const path = require('path');
```

File: rules/S2083/javascript/how-to-fix-it/node.adoc [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+
+function (req, res) {
+  const userFilename = fs.realPath(__dirname + req.query.filename);
+
+  const targetDirectory = __dirname + '/User/John';
+  if (!realPath.startsWith(targetDirectory)) {
```
Comment (loris-s-sonarsource) [inline rules/S2083/javascript/how-to-fix-it/node.adoc:57]:
```suggestion
  const resourcesDir = "/data/app/resources/"
  const userFilename = path.join(resourcesDir, req.query.filename));
  const userFilename = fs.realPath(userFilename);

  if (!userFilename.startsWith(resourcesDir)) {
```

File: rules/S2083/javascript/how-to-fix-it/node.adoc (hunk 5 above)
Comment (loris-s-sonarsource) [reply rules/S2083/javascript/how-to-fix-it/node.adoc:56]: done in another comment

File: rules/S2083/javascript/how-to-fix-it/node.adoc [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+    res.status(401).send();
+  }
+
+  let data = fs.readFileSync(userFilename, { encoding: 'utf8', flag: 'r' });
+}
+----
+
+=== How does this work?
+
+include::../../common/fix/how-does-this-work.adoc[]
```
Comment (loris-s-sonarsource) [inline rules/S2083/javascript/how-to-fix-it/node.adoc:43]:
```suggestion
include::../../common/fix/self-validation.adoc[]
```

File: rules/S2083/javascript/how-to-fix-it/node.adoc [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+
+=== How does this work?
+
+include::../../common/fix/self-validation.adoc[]
+
+*Important Note*: The order of this process pattern is important. The code must follow this order exactly to be secure by design:
+
+1. `data = transform(user_input);`
+2. `data = normalize(data);`
+3. `data = clean(data);`
```
Comment (loris-s-sonarsource) [inline rules/S2083/javascript/how-to-fix-it/node.adoc:49]:
```suggestion
3. `data = sanitize(data);`
```

File: rules/S2083/javascript/how-to-fix-it/node.adoc [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+4. `use(data);`
+
+:tnsu_talk: https://www.youtube.com/watch?v=V-DdcKADnFk
+As pointed out in {tnsu_talk}[this SonarSource talk], failure to follow this exact order undoubtedly leads to security vulnerabilities.
```
Comment (loris-s-sonarsource) [inline rules/S2083/javascript/how-to-fix-it/node.adoc:53]:
```suggestion
As pointed out in {tnsu_talk}[this SonarSource talk], failure to follow this exact order leads to security vulnerabilities.
```

File: rules/S2083/javascript/how-to-fix-it/express.adoc [hunk 14]
```diff
@@ -0,0 +1,46 @@
+=== How to fix it in Express.Js
+
+The following code is vulnerable to path injection as it is constructing a path
+using untrusted data. +
+This path is then used without validation to return a file.
+
+Therefore, it can be leveraged by an attacker to read arbitrary files.
+
+==== Non-compliant code example
```
Comment (loris-s-sonarsource) [inline rules/S2083/javascript/how-to-fix-it/express.adoc:9]:
```suggestion
==== Noncompliant code example
```

File: rules/S2083/javascript/how-to-fix-it/node.adoc [hunk 15]
```diff
@@ -0,0 +1,89 @@
+=== How to fix it in Node.js
+
+:canonicalization_function: `fs.realPath`
+include::../../common/fix/code-rationale.adoc[]
+
+==== Non-compliant code example
```
Comment (loris-s-sonarsource) [inline rules/S2083/javascript/how-to-fix-it/node.adoc:6]:
```suggestion
==== Noncompliant code example
```

File: rules/S2083/javascript/how-to-fix-it/node.adoc [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+
+*Important Note*: The order of this process pattern is important. The code must follow this order exactly to be secure by design:
+
+1. `data = transform(user_input);`
+2. `data = normalize(data);`
+3. `data = sanitize(data);`
+4. `use(data);`
+
+:tnsu_talk: https://www.youtube.com/watch?v=V-DdcKADnFk
+As pointed out in {tnsu_talk}[this SonarSource talk], failure to follow this exact order leads to security vulnerabilities.
```
Comment (loris-s-sonarsource) [inline rules/S2083/javascript/how-to-fix-it/node.adoc:53]:
```suggestion
```

File: rules/S2083/javascript/how-to-fix-it/node.adoc [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+  }
+
+  let data = fs.readFileSync(userFilename, { encoding: 'utf8', flag: 'r' });
+}
+----
+
+=== How does this work?
+
+include::../../common/fix/self-validation.adoc[]
+
```
Comment (loris-s-sonarsource) [inline rules/S2083/javascript/how-to-fix-it/node.adoc:44]: See https://github.com/SonarSource/rspec/pull/1328/files#diff-26e00bf0bb3adaf085510cde100874b7d349084c9070373acb2e848d08e6e0a8

File: rules/S2083/javascript/how-to-fix-it/express.adoc [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+==== Compliant solution
+
+[source,javascript,diff-id=1,diff-type=compliant]
+----
+const fs = require('fs');
+const path = require('path');
+
+function (req, res) {
+  const targetDirectory = "/data/app/resources/";
+  const userFilename = path.join(resourcesDir, req.query.filename);
```
Comment (gaetan-ferry-sonarsource) [inline rules/S2083/javascript/how-to-fix-it/express.adoc:33]: `resourcesDir` is not defined here

File: rules/S2083/javascript/how-to-fix-it/express.adoc [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+
+  const targetDirectory = process.cwd();
```
Comment (gaetan-ferry-sonarsource) [inline rules/S2083/javascript/how-to-fix-it/express.adoc:35]: Apparently, there is a mixup with the `targetDirectory` / `resourcesDir names` I would suggest:
```
const targetDirectory = "/data/app/resources/";
const userFilename = path.join(targetDirectory, req.query.filename);
res.sendFile(userFilename, { root: targetDirectory });
```

File: rules/S2083/javascript/how-to-fix-it/node.adoc [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+that the string `targetDirectory` does not end with a path separator:
+
+
+[source, javascript]
+----
+const fs = require('fs');
+const path = require('path');
+
+function (req, res) {
+  const resourcesDir = "/data/app/resources/"
```
Comment (gaetan-ferry-sonarsource) [inline rules/S2083/javascript/how-to-fix-it/node.adoc:59]: I don't think it is vulnerable if you add the forward slash here.
```suggestion
  const resourcesDir = "/data/app/resources"
```

**Review by gaetan-ferry-sonarsource — CHANGES_REQUESTED** [review]
LGTM in general. However, there are still a few inconsistencies in the code samples that need to be fixed.

File: rules/S2083/javascript/how-to-fix-it/express.adoc (hunk 18 above)
Comment (loris-s-sonarsource) [inline rules/S2083/javascript/how-to-fix-it/express.adoc:33]:
```suggestion
  const userFilename = path.join(targetDirectory, req.query.filename);
```

File: rules/S2083/javascript/how-to-fix-it/express.adoc [hunk 21]
```diff
@@ -0,0 +1,46 @@
...
+==== Noncompliant code example
+
+[source,javascript,diff-id=1,diff-type=noncompliant]
+----
+const fs = require('fs');
+const path = require('path');
+
+function (req, res) {
+  const resourcesDir = "/data/app/resources/"
+  const userFilename = path.join(resourcesDir, req.query.filename);
```
Comment (loris-s-sonarsource) [inline rules/S2083/javascript/how-to-fix-it/express.adoc:18]:
```suggestion
  const targetDirectory = "/data/app/resources/"
  const userFilename = path.join(targetDirectory, req.query.filename);
```

File: rules/S2083/javascript/how-to-fix-it/express.adoc (hunk 19 above)
Comment (loris-s-sonarsource) [reply rules/S2083/javascript/how-to-fix-it/express.adoc:35]:
```suggestion
```

File: rules/S2083/javascript/how-to-fix-it/express.adoc [hunk 22]
```diff
@@ -0,0 +1,45 @@
...
+==== Compliant solution
+
+[source,javascript,diff-id=1,diff-type=compliant]
+----
+const fs = require('fs');
+const path = require('path');
+
+function (req, res) {
+  const targetDirectory = "/data/app/resources/";
+  const userFilename = path.join(targetDirectory, req.query.filename);
```
Comment (loris-s-sonarsource) [inline rules/S2083/javascript/how-to-fix-it/express.adoc:33]:
```suggestion
  const userFilename = req.query.filename;
```

File: rules/S2083/javascript/how-to-fix-it/node.adoc [hunk 23]
```diff
@@ -0,0 +1,79 @@
...
+==== Noncompliant code example
+
+[source,javascript,diff-id=1,diff-type=noncompliant]
+----
+const fs = require('fs');
+const path = require('path');
+
+function (req, res) {
+  const resourcesDir = "/data/app/resources/"
+  const userFilename = path.join(resourcesDir, req.query.filename);
```
Comment (loris-s-sonarsource) [inline rules/S2083/javascript/how-to-fix-it/node.adoc:15]:
```suggestion
  const targetDirectory = "/data/app/resources/"
  const userFilename = path.join(targetDirectory, req.query.filename);
```

File: rules/S2083/javascript/how-to-fix-it/node.adoc [hunk 24]
```diff
@@ -0,0 +1,79 @@
...
+----
+const fs = require('fs');
+const path = require('path');
+
+function (req, res) {
+  const resourcesDir = "/data/app/resources/"
+  const userFilename = path.join(resourcesDir, req.query.filename));
+  const userFilename = fs.realPath(userFilename);
+
+  if (!userFilename.startsWith(resourcesDir)) {
```
Comment (loris-s-sonarsource) [inline rules/S2083/javascript/how-to-fix-it/node.adoc:33]:
```suggestion
  const targetDirectory = "/data/app/resources/"
  const userFilename = path.join(targetDirectory, req.query.filename));
  const userFilename = fs.realPath(userFilename);

  if (!userFilename.startsWith(resourcesDir)) {
```

File: rules/S2083/javascript/how-to-fix-it/node.adoc [hunk 25]
```diff
@@ -0,0 +1,79 @@
...
+----
+const fs = require('fs');
+const path = require('path');
+
+function (req, res) {
+  const resourcesDir = "/data/app/resources"
+  const userFilename = path.join(resourcesDir, req.query.filename));
+  const userFilename = fs.realPath(userFilename);
+
+  if (!userFilename.startsWith(resourcesDir)) {
```
Comment (loris-s-sonarsource) [inline rules/S2083/javascript/how-to-fix-it/node.adoc:63]:
```suggestion
  const targetDirectory = "/data/app/resources"
  const userFilename = path.join(targetDirectory, req.query.filename));
  const userFilename = fs.realPath(userFilename);

  if (!userFilename.startsWith(targetDirectory)) {
```

File: rules/S2083/javascript/how-to-fix-it/express.adoc [hunk 26]
```diff
@@ -0,0 +1,45 @@
...
+
+  res.sendFile(userFilename); // Noncompliant
+}
+----
+
+==== Compliant solution
+
+[source,javascript,diff-id=1,diff-type=compliant]
+----
+const fs = require('fs');
```
Comment (loris-s-sonarsource) [inline rules/S2083/javascript/how-to-fix-it/express.adoc:28]:
```suggestion
```

File: rules/S2083/javascript/how-to-fix-it/node.adoc [hunk 27]
```diff
@@ -0,0 +1,79 @@
+=== How to fix it in Node.js
+
+:canonicalization_function: `fs.realPath`
+include::../../common/fix/code-rationale.adoc[]
+
+==== Noncompliant code example
+
+[source,javascript,diff-id=1,diff-type=noncompliant]
+----
+const fs = require('fs');
```
Comment (loris-s-sonarsource) [inline rules/S2083/javascript/how-to-fix-it/node.adoc:10]:
```suggestion
```

File: rules/S2083/javascript/how-to-fix-it/node.adoc [hunk 28]
```diff
@@ -0,0 +1,79 @@
...
+
+include::../../common/pitfalls/partial-path-traversal.adoc[]
+
+For example, the following code is vulnerable to partial path injection. Note
+that the string `targetDirectory` does not end with a path separator:
+
+
+[source, javascript]
+----
+const fs = require('fs');
```
Comment (loris-s-sonarsource) [inline rules/S2083/javascript/how-to-fix-it/node.adoc:55]:
```suggestion
```

File: rules/S2083/javascript/how-to-fix-it/express.adoc [hunk 29]
```diff
@@ -0,0 +1,44 @@
...
+using untrusted data. +
+This path is then used without validation to return a file.
+
+Therefore, it can be leveraged by an attacker to read arbitrary files.
+
+==== Noncompliant code example
+
+[source,javascript,diff-id=1,diff-type=noncompliant]
+----
+const fs = require('fs');
```
Comment (loris-s-sonarsource) [inline rules/S2083/javascript/how-to-fix-it/express.adoc:13]:
```suggestion
```

**Review by gaetan-ferry-sonarsource — APPROVED** [review]
LGTM

**Pull request merged**